sherpa/models/model.py

```diff
@@ -731,6 +731,75 @@ def __setattr__(self, name: str, val) -> None:
         for alias in val.aliases:
             self._par_index[alias] = val
 
+    def get_parts(self,
+                  include_composites : bool = True,
+                  remove_duplicates : bool = False) -> Sequence[Model]:
```

Comment:
```suggestion
                 remove_duplicates : bool = False) -> list[Model]:
```

There's a whole thing about contravariant vs covariant types, and I think we want to be explicit in the return type rather than a more-generic type.

Reply:
oops - missed a space.

Reply:
I'm now to typing (I think this is the first PR where I try to ype things like that) and mypy as complaining about exactly contravariant vs covariant types. I have ot admit I don't understand it and it also doesn't seem to matter much as a list is just a specialized Sequence, so I figured I just take whatever makes the warning go away. But then it shows up the other way around somewhere else. Maybe it's better ot just NOT type is in that case?

Reply:
So, the typing docuementation doesn't explicitly call this out at the moment, but you have statements like

> To summarize, function parameters should behave contravariantly and function return types should behave covariantly.

that occur **just before** this link: https://typing.python.org/en/latest/spec/callables.html#passing-kwargs-inside-a-function-to-another-function

Or this may be better: from https://typing.python.org/en/latest/spec/callables.html#parameter-types

> Callable types are covariant with respect to their return types but contravariant with respect to their parameter types.

[this is all about using `Callable` but I assume it somewhat holds elsewhere].

The `mypy` documentation can also be looked at: from https://mypy.readthedocs.io/en/stable/generics.html#variance-of-generics includes

> Most immutable container types, such as [Sequence](https://docs.python.org/3/library/collections.abc.html#collections.abc.Sequence) and [frozenset](https://docs.python.org/3/library/stdtypes.html#frozenset) are covariant.

So this suggests that the way @hamogu had it originally made sense - but then all the other `get_parts` calls need to be converted to `...) -> Sequence[Model]`. 

I had thought I'd started to get a handle on this but I can no-longer find the documentation I had been following.

Reply:
Thank you for the explanation, but I think I figured it out with an extra typing hint in the middle of the code: `parts : list[Model] = []`. I just needed to tell MyPy what `parts` is supposed to be so it knows that it's not just "a general list" but "a list with Models".